entities/sync/sema.go

```diff
@@ -0,0 +1,45 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2024 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package sync
+
+import (
+	"context"
+
+	"golang.org/x/sync/semaphore"
+)
+
+// Semaphore is a wrapper around a semaphore implementation to have a zero-value Semaphore which is a noop.
+type Semaphore struct {
+	sema *semaphore.Weighted
+}
+
+func NewSemaphore(n int) Semaphore {
+	return Semaphore{
+		sema: semaphore.NewWeighted(int64(n)),
+	}
+}
+
+func (s Semaphore) Acquire(ctx context.Context) error {
+	if s.sema == nil {
```

Comment:
silent noop may be dangerous as using a not initialized semaphore would give a false controlled situation

Reply:
this was supposed to be a feature, but probably yeah. I will remove this class altogether and make the semaphore a requirement